File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def dfs(node):
            if node and node.val not in adrs:
                n=Node(node.val)
                adrs[n.val]=n
                for i in node.neighbors:
                    dfs(i)
        
        def dfs2(node):
            if len(adrs[node.val].neighbors) != len(node.neighbors):
                for i in node.neighbors:
                    adrs[node.val].neighbors.append(adrs[i.val])
                for i in node.neighbors:
                    dfs2(i)
                

        if node:
            adrs={}
            dfs(node)
            dfs2(node)
            return adrs[node.val]

File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    Node[] vis;
    public Node cloneGraph(Node node) {
        if(node==null) return null;
        vis = new Node[101];
        vis[node.val] = new Node(node.val);
        return dfs(node, vis[node.val]);
    }
    
    Node dfs(Node node, Node ans){
        
        for(Node nbr: node.neighbors){
            if(vis[nbr.val]==null){
                vis[nbr.val] = new Node(nbr.val);
                ans.neighbors.add(vis[nbr.val]);
                dfs(nbr, vis[nbr.val]);
            }
            else
            ans.neighbors.add(vis[nbr.val]);
        }
        
        return ans;
    }
}